cranelift/codegen/src/alias_analysis.rs

```diff
@@ -0,0 +1,387 @@
+            mem_values: FxHashMap::default(),
+        };
+
+        analysis.compute_block_input_states();
+        analysis
+    }
+
+    fn compute_block_input_states(&mut self) {
+        let mut queue = vec![];
+        let mut queue_set = FxHashSet::default();
+        let entry = self.func.layout.entry_block().unwrap();
+        queue.push(entry);
+        queue_set.insert(entry);
+
+        while let Some(block) = queue.pop() {
+            queue_set.remove(&block);
+            let mut state = self
+                .block_input
+                .entry(block)
+                .or_insert_with(|| LastStores::default())
+                .clone();
+
+            log::trace!(
+                "alias analysis: input to block{} is {:?}",
+                block.index(),
+                state
+            );
+
+            for inst in self.func.layout.block_insts(block) {
+                state.update(self.func, inst);
+                log::trace!("after inst{}: state is {:?}", inst.index(), state);
+            }
+
+            visit_block_succs(self.func, block, |_inst, succ| {
+                let succ_first_inst = self
+                    .func
+                    .layout
+                    .block_insts(succ)
+                    .into_iter()
+                    .next()
+                    .unwrap();
+                let updated = match self.block_input.get_mut(&succ) {
+                    Some(succ_state) => {
+                        let old = succ_state.clone();
+                        succ_state.meet_from(&state, succ_first_inst);
+                        *succ_state != old
+                    }
+                    None => {
+                        self.block_input.insert(succ, state.clone());
+                        true
+                    }
+                };
+
+                if updated && queue_set.insert(succ) {
+                    queue.push(succ);
+                }
+            });
+        }
+    }
+
+    /// Make a pass and update known-redundant loads to aliased
+    /// values. We interleave the updates with the memory-location
+    /// tracking because resolving some aliases may expose others
+    /// (e.g. in cases of double-indirection with two separate chains
+    /// of loads).
+    pub fn compute_and_update_aliases(&mut self) {
+        let mut pos = FuncCursor::new(self.func);
+
+        while let Some(block) = pos.next_block() {
+            let mut state = self
+                .block_input
+                .get(&block)
+                .cloned()
+                .unwrap_or_else(|| LastStores::default());
+
+            while let Some(inst) = pos.next_inst() {
+                log::trace!(
+                    "alias analysis: scanning at inst{} with state {:?} ({:?})",
+                    inst.index(),
+                    state,
+                    pos.func.dfg[inst],
+                );
+
+                if let Some((address, offset, ty)) = inst_addr_offset_type(pos.func, inst) {
```

Comment:
It might make sense to reduce indentation here, since this is basically the rest of the function:

```rust
let (address, offset, ty) = match inst_addr_offset_type(pos.func, inst) {
    None => continue,
    Some(x) => x,
};

// ...
```

Totally up to you.

Reply:
Ah, there's actually a final call to `update_state` below the long if-let, which this would skip, and factoring the body out is a bit complex because of mutable borrows of self. Left it as-is for now, but thanks in any case!